File: apps/macos/ElyanMac/TaskInboxView.swift
import SwiftUI

/// Shows tasks the mobile app has dispatched to this desktop. Assignment is
/// SSE-pushed (`AppState.refreshAssignedTasks`, triggered by task/command/
/// runtime realtime events) rather than polled; the Python runtime is the
/// sole executor when operational (`supervisor.executeAssignedTasks()`) and
/// Swift only auto-acks as a degraded fallback when the runtime isn't
/// available (see AppState.refreshAssignedTasks). Runtime completion is
/// evidence-driven; the UI can cancel or approve but cannot fabricate a
/// completed state. A task the runtime marked
/// `waiting_approval` gets a real approve/deny sheet wired to
/// `supervisor.approveTask` instead.
struct TaskInboxView: View {
    @EnvironmentObject var appState: AppState
    @State private var approvalItem: RuntimeTaskItem?

    var body: some View {
        VStack(spacing: 0) {
            header
            Divider()
            if appState.assignedTasks.isEmpty {
                emptyState
            } else {
                list
            }
        }
        .background(Color(NSColor.windowBackgroundColor))
        .task { await appState.refreshAssignedTasks() }
        .sheet(item: $approvalItem) { item in
            TaskApprovalView(
                task: item,
                onApprove: {
                    Task {
                        await appState.supervisor.approveTask(taskId: item.id, approved: true)
                        approvalItem = nil
                        await appState.refreshAssignedTasks()
                    }
                },
                onDeny: {
                    Task {
                        await appState.supervisor.approveTask(taskId: item.id, approved: false)
                        approvalItem = nil
                        await appState.refreshAssignedTasks()
                    }
                }
            )
        }
    }

    /// The Python bridge's own task inbox is the authoritative source for
    /// `approvalRequest` payloads (title/message/labels) — the REST
    /// `ElyanRuntimeTask` model the list above renders from doesn't carry
    /// that detail. Match by id to find the matching bridge-side record.
    private func approvalItem(for task: ElyanRuntimeTask) -> RuntimeTaskItem? {
        guard task.status.lowercased() == "waiting_approval" else { return nil }
        return appState.supervisor.taskInbox.first { $0.id == task.id && $0.isWaitingApproval }
    }

    private var header: some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                Text("Mobilden görevler")
                    .font(.title3.bold())
                Text("Telefondan gönderilen işler burada listelenir.")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
            Spacer()
            Button {
                Task { await appState.refreshAssignedTasks() }
            } label: {
                Image(systemName: "arrow.clockwise")
            }
            .buttonStyle(.plain)
            .foregroundStyle(.secondary)
        }
        .padding(20)
    }

    private var emptyState: some View {
        VStack(spacing: 10) {
            Spacer(minLength: 40)
            Image(systemName: "tray")
                .font(.system(size: 34, weight: .light))
                .foregroundStyle(.tertiary)
            Text("Henüz görev yok")
                .font(.headline)
            Text("Mobilde bir görev gönderdiğinde burada görürsün.")
                .font(.callout)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
                .frame(maxWidth: 320)
            if !appState.lastDispatchError.isEmpty {
                Text(appState.lastDispatchError)
                    .font(.caption)
                    .foregroundStyle(.red)
                    .padding(.top, 8)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: 360)
            }
            Spacer()
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(24)
    }

    private var list: some View {
        ScrollView {
            LazyVStack(spacing: 12) {
                ForEach(appState.assignedTasks) { task in
                    TaskRow(task: task,
                            pendingApproval: approvalItem(for: task),
                            onReview: { item in
                                approvalItem = item
                            })
                }
            }
            .padding(20)
        }
    }
}

private struct TaskRow: View {
    let task: ElyanRuntimeTask
    var pendingApproval: RuntimeTaskItem? = nil
    var onReview: (RuntimeTaskItem) -> Void = { _ in }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(alignment: .top, spacing: 10) {
                Circle()
                    .fill(statusColor.opacity(0.2))
                    .frame(width: 10, height: 10)
                    .padding(.top, 6)
                VStack(alignment: .leading, spacing: 4) {
                    Text(task.title)
                        .font(.system(size: 14, weight: .semibold))
                    if !task.summary.isEmpty {
                        Text(task.summary)
                            .font(.system(size: 12))
                            .foregroundStyle(.secondary)
                    }
                    Text(statusLabel)
                        .font(.system(size: 11, weight: .medium))
                        .foregroundStyle(statusColor)
                }
                Spacer()
            }
            if !task.error.isEmpty {
                Text(task.error)
                    .font(.system(size: 11))
                    .foregroundStyle(.red)
            }
            if let pendingApproval {
                HStack(spacing: 8) {
                    Spacer()
                    Button("İncele ve Onayla") { onReview(pendingApproval) }
                        .buttonStyle(.borderedProminent)
                }
            }
        }
        .padding(14)
        .background(Color(NSColor.controlBackgroundColor).opacity(0.6))
        .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: 10, style: .continuous)
                .strokeBorder(Color.primary.opacity(0.08), lineWidth: 1)
        )
    }

    private var statusColor: Color {
        switch task.status.lowercased() {
        case "completed": return .green
        case "failed", "canceled": return .red
        case "waiting_approval": return .orange
        default: return .accentColor
        }
    }

    private var statusLabel: String {
        switch task.status.lowercased() {
        case "queued": return "Sırada"
        case "running", "planning": return "Çalışıyor"
        case "waiting_approval": return "Onay bekliyor"
        case "completed": return "Tamamlandı"
        case "failed": return "Başarısız"
        case "canceled": return "İptal edildi"
        default: return task.status.capitalized
        }
    }
}

File: apps/desktop/lib/ui/task_inbox_view.dart
import 'package:flutter/material.dart';

import '../models/chat_models.dart';
import '../store/app_state.dart';
import '../theme/elyan_theme.dart';

/// Mobilden gönderilen görevler — macOS `TaskInboxView.swift` portu.
/// Yürütücü her zaman Python runtime'dır; bu ekran görünürlük + onay
/// (waiting_approval → İncele ve Onayla) içindir.
class TaskInboxView extends StatefulWidget {
  final AppState appState;
  const TaskInboxView({super.key, required this.appState});

  @override
  State<TaskInboxView> createState() => _TaskInboxViewState();
}

class _TaskInboxViewState extends State<TaskInboxView> {
  bool _isRefreshing = false;

  @override
  void initState() {
    super.initState();
    _refresh();
  }

  Future<void> _refresh() async {
    if (_isRefreshing) return;
    setState(() => _isRefreshing = true);
    await widget.appState.supervisor.refreshTaskInbox();
    if (mounted) setState(() => _isRefreshing = false);
  }

  Future<void> _decide(RuntimeTaskItem task, bool approved) async {
    Navigator.of(context).pop();
    await widget.appState.supervisor
        .approveTask(taskId: task.id, approved: approved);
    if (mounted) setState(() {});
  }

  void _showApprovalSheet(RuntimeTaskItem task) {
    final approval = task.approvalRequest;
    if (approval == null) return;
    final brightness = Theme.of(context).brightness;
    showDialog<void>(
      context: context,
      builder: (context) => AlertDialog(
        backgroundColor: ElyanTheme.surface(brightness),
        shape:
            RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
        title: Text(approval.title,
            style:
                const TextStyle(fontSize: 16, fontWeight: FontWeight.w600)),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            if (approval.message.isNotEmpty)
              Text(approval.message, style: const TextStyle(fontSize: 13)),
            if (approval.summary.isNotEmpty) ...[
              const SizedBox(height: 8),
              Text(
                approval.summary,
                style: TextStyle(
                  fontSize: 12,
                  color: ElyanTheme.secondaryText(brightness),
                ),
              ),
            ],
          ],
        ),
        actions: [
          OutlinedButton(
            onPressed: () => _decide(task, false),
            child: Text(approval.rejectLabel),
          ),
          FilledButton(
            style: FilledButton.styleFrom(backgroundColor: ElyanTheme.accent),
            onPressed: () => _decide(task, true),
            child: Text(approval.confirmLabel),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final brightness = Theme.of(context).brightness;
    final tasks = widget.appState.supervisor.taskInbox;
    return Container(
      color: ElyanTheme.canvas(brightness),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Padding(
            padding: const EdgeInsets.all(20),
            child: Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      const Text(
                        'Mobilden görevler',
                        style: TextStyle(
                            fontSize: 17, fontWeight: FontWeight.bold),
                      ),
                      const SizedBox(height: 2),
                      Text(
                        'Telefondan gönderilen işler burada listelenir.',
                        style: TextStyle(
                          fontSize: 12,
                          color: ElyanTheme.secondaryText(brightness),
                        ),
                      ),
                    ],
                  ),
                ),
                IconButton(
                  onPressed: _refresh,
                  icon: _isRefreshing
                      ? const SizedBox(
                          width: 14,
                          height: 14,
                          child: CircularProgressIndicator(strokeWidth: 2),
                        )
                      : Icon(Icons.refresh,
                          size: 18,
                          color: ElyanTheme.secondaryText(brightness)),
                ),
              ],
            ),
          ),
          Container(height: 1, color: ElyanTheme.hairline(brightness)),
          Expanded(
            child: tasks.isEmpty
                ? _emptyState(brightness)
                : ListView(
                    padding: const EdgeInsets.all(20),
                    children: [
                      for (final task in tasks) ...[
                        _TaskRow(
                          task: task,
                          brightness: brightness,
                          onReview: () => _showApprovalSheet(task),
                        ),
                        const SizedBox(height: 12),
                      ],
                    ],
                  ),
          ),
        ],
      ),
    );
  }

  Widget _emptyState(Brightness brightness) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(Icons.inbox_outlined,
              size: 34, color: ElyanTheme.tertiaryText(brightness)),
          const SizedBox(height: 10),
          const Text('Henüz görev yok',
              style: TextStyle(fontSize: 14, fontWeight: FontWeight.w600)),
          const SizedBox(height: 4),
          SizedBox(
            width: 320,
            child: Text(
              'Mobilde bir görev gönderdiğinde burada görürsün.',
              textAlign: TextAlign.center,
              style: TextStyle(
                fontSize: 13,
                color: ElyanTheme.secondaryText(brightness),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _TaskRow extends StatelessWidget {
  final RuntimeTaskItem task;
  final Brightness brightness;
  final VoidCallback onReview;

  const _TaskRow({
    required this.task,
    required this.brightness,
    required this.onReview,
  });

  Color get _statusColor {
    switch (task.status) {
      case 'completed':
        return Colors.green;
      case 'failed':
      case 'canceled':
        return Colors.red;
      case 'waiting_approval':
        return Colors.orange;
      default:
        return ElyanTheme.accent;
    }
  }

  String get _statusLabel {
    switch (task.status) {
      case 'queued':
        return 'Sırada';
      case 'running':
      case 'planning':
        return 'Çalışıyor';
      case 'waiting_approval':
        return 'Onay bekliyor';
      case 'completed':
        return 'Tamamlandı';
      case 'failed':
        return 'Başarısız';
      case 'canceled':
        return 'İptal edildi';
      default:
        return task.status;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.all(14),
      decoration: BoxDecoration(
        color: ElyanTheme.surface(brightness),
        borderRadius: BorderRadius.circular(10),
        border: Border.all(color: ElyanTheme.hairline(brightness)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Padding(
                padding: const EdgeInsets.only(top: 6),
                child: Container(
                  width: 10,
                  height: 10,
                  decoration: BoxDecoration(
                    color: _statusColor.withValues(alpha: 0.2),
                    shape: BoxShape.circle,
                  ),
                ),
              ),
              const SizedBox(width: 10),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(task.title,
                        style: const TextStyle(
                            fontSize: 14, fontWeight: FontWeight.w600)),
                    if (task.summary.isNotEmpty) ...[
                      const SizedBox(height: 4),
                      Text(
                        task.summary,
                        style: TextStyle(
                          fontSize: 12,
                          color: ElyanTheme.secondaryText(brightness),
                        ),
                      ),
                    ],
                    const SizedBox(height: 4),
                    Text(
                      _statusLabel,
                      style: TextStyle(
                        fontSize: 11,
                        fontWeight: FontWeight.w500,
                        color: _statusColor,
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
          if (task.error.isNotEmpty) ...[
            const SizedBox(height: 8),
            Text(task.error,
                style: const TextStyle(fontSize: 11, color: Colors.red)),
          ],
          if (task.isWaitingApproval) ...[
            const SizedBox(height: 8),
            Align(
              alignment: Alignment.centerRight,
              child: FilledButton(
                style: FilledButton.styleFrom(
                    backgroundColor: ElyanTheme.accent),
                onPressed: onReview,
                child: const Text('İncele ve Onayla',
                    style: TextStyle(fontSize: 12)),
              ),
            ),
          ],
        ],
      ),
    );
  }
}
